src/compiler/irgen.jl

```diff
@@ -278,10 +373,196 @@ function emit_exception!(builder, name, inst)
+                            if succ != bb
+                                fallthrough = succ
+                                break
+                            end
+                        end
+                        fallthrough === nothing || break
+
+                        # recurse upwards
+                        old_preds = copy(preds)
+                        empty!(preds)
+                        for pred in old_preds
+                            append!(preds, predecessors[pred])
+                        end
+                    end
+                    push!(worklist, bb => fallthrough)
+
+                    dispose(builder)
+                end
+            end
+        end
+    end
+
+    # apply the pending terminator rewrites
+    @timeit to[] "replace" if !isempty(worklist)
+        let builder = Builder(JuliaContext())
+            for (bb, fallthrough) in worklist
+                position!(builder, bb)
+                if fallthrough !== nothing
+                    br!(builder, fallthrough)
+                else
+                    # couldn't find any other successor. this happens with functions
+                    # that only contain a single block, or when the block is dead.
+                    ft = eltype(llvmtype(fun))
+                    if return_type(ft) == LLVM.VoidType(JuliaContext())
+                        # even though returning can lead to invalid control flow,
+                        # it mostly happens with functions that just throw,
+                        # and leaving the unreachable there would make the optimizer
+                        # place another after the call.
+                        ret!(builder)
+                    else
+                        unreachable!(builder)
+                    end
+                end
+            end
+        end
+    end
+
+    end
+    return changed
+end
+
+# HACK: this pass removes calls to `trap` and replaces them with inline assembly
+#
+# if LLVM knows we're trapping, code is marked `unreachable` (see `hide_unreachable!`).
+function hide_trap!(mod::LLVM.Module)
+    job = current_job::CompilerJob
+    changed = false
+    @timeit to[] "hide trap" begin
+
+    # inline assembly to exit a thread, hiding control flow from LLVM
+    exit_ft = LLVM.FunctionType(LLVM.VoidType(JuliaContext()))
+    exit = InlineAsm(exit_ft, "exit;", "", true)
+
+    if haskey(functions(mod), "llvm.trap")
+        trap = functions(mod)["llvm.trap"]
+
+        for use in uses(trap)
+            val = user(use)
+            if isa(val, LLVM.CallInst)
+                let builder = Builder(JuliaContext())
+                    position!(builder, val)
+                    call!(builder, exit)
+                    dispose(builder)
+                end
+                unsafe_delete!(LLVM.parent(val), val)
+                changed = true
+            end
+        end
+    end
+
+    end
+    return changed
+end
+
+# emit a global variable for storing the current exception status
```

Comment:
I'm not sure how exceptions should be done in AMDGCN, since I don't believe we have the same facility that CUDA does for this. Maybe we can implement our own? @vchuravy what do you think we should do here?

Reply:
Let's save this for a future PR